banditpylib/learners/collaborative_learner/collaborative_learning.py

```diff
@@ -0,0 +1,390 @@
+
+    arm_pull = actions.arm_pulls.add()
+
+    # map local arm index to the bandits arm index
+    arm_pull.arm.id = self.__assigned_arms[int(np.argmax(self.__ucb()))]
+    arm_pull.times = 1
+
+    return actions
+
+  def update(self, feedback: Feedback):
+    for arm_feedback in feedback.arm_feedbacks:
+      # reverse map from bandit index to local index
+      pseudo_arm_index = np.where(
+        self.__assigned_arms==arm_feedback.arm.id)[0][0]
+      self.__pseudo_arms[pseudo_arm_index].update(
+          np.array(arm_feedback.rewards))
+      self.__total_pulls += len(arm_feedback.rewards)
+
+    if self.__stage == 'initialization':
+      self.__stage = 'main'
+
+  @property
+  def best_arm(self) -> int:
+    # map best arm local index to actual bandit index
+    return self.__assigned_arms[
+      argmax_or_min_tuple([
+        (pseudo_arm.total_pulls, arm_id)
+        for (arm_id, pseudo_arm) in enumerate(self.__pseudo_arms)
+      ])
+    ]
+
+  def get_total_pulls(self) -> int:
+    return self.__total_pulls
+
+
+class LilUCBHeuristicCollaborativeBAIAgent(CollaborativeBAIAgent):
+  r"""Implementation of agent of the Collaborative Learning Algorithm\
+  Uses LilUCBHeuristic as the central algorithm
+
+  :param int arm_num: number of arms of the bandit
+  :param int rounds: number of rounds of communication allowed
+    (this agent uses one more)
+  :param int horizon: maximum number of pulls the agent can make
+    (over all rounds combined)
+  :param Optional[str] name: alias name
+  """
+
+  def __init__(self, arm_num: int, rounds: int,
+    horizon: int, name: Optional[str] = None):
+    super().__init__(name)
+    if arm_num <= 1:
+      raise ValueError('Number of arms is expected at least 2. Got %d.' %
+                       arm_num)
+    if rounds <= 1:
+      raise ValueError('Number of rounds is expected at least 2. Got %d.' %
+                       rounds)
+    self.__arm_num = arm_num
+    self.__rounds = rounds - 1
+    self.__horizon = horizon
+    self.__num_pulls_learning = int(0.5 * horizon / self.__rounds)
+
+  def _name(self) -> str:
+    return "lilUCBHeur_collaborative_agent"
+
+  def reset(self):
+    self.__round_num = 0
+    self.__stage = "unassigned"
+    # True if action forwarded from central algo
+    self.__central_algo_action_taken = False
+
+  def __complete_round(self):
+    self.__round_num += 1
+    self.__central_algo_action_taken = False
+    if self.__round_num < self.__rounds + 1:
+      self.__stage = "unassigned"
+    else:
+      self.__stage = "termination"
+
+  def set_input_arms(self, arms: List[int]):
+    if not arms:
+      # terminate if no arms assigned
+      self.__stage = "termination"
+      return
+
+    self.__assigned_arms = np.array(arms)
+    # confidence of 0.01 suggested in the paper
+    self.__central_algo = LilUCBHeuristicCollaborative(self.__arm_num,
+      0.99, self.__assigned_arms)
+    self.__central_algo.reset()
+    self.__stage = "preparation"
```

Comment:
Why in set_input_arms, we always set the stage to `preparation`?

Reply:
setting input arms only occurs before every preparation step

Reply:
I have changed it to 
```
if self.__stage == "unassigned":
      self.__stage = "preparation"
```